compiler/modules/front.syntax/preparseStmt.bal

```
@@ -1,41 +1,38 @@
 import wso2/nballerina.err;
 
-const PREDICT_TYPE_DESC = true;
-const PREDICT_EXPR = false;
-
+const PREPARSE_TYPE_DESC = true;
+const PREPARSE_EXPR = false;
 
 // Returns `true` if a statement whose first token was `(` begins a type descriptor rather than an expression
 // (implying that the statement is a local variable declaration rather than a method call).
 // This is a preparse: the statement will be parsed again according to the value returned.
-// For now, this only handles a statement starting with `()`. It will be expanded later.
-function preparseParenTypeDesc(Tokenizer tok) returns boolean|err:Syntax {
+function preparseParenTypeDesc(Tokenizer tok, string? close) returns boolean|err:Syntax {
+    boolean? result = check skipToClose(tok, close);
     Token? t = tok.current();
-    if t != ")" {
-        return PREDICT_EXPR;
+    while t == "[" {
+        check tok.advance();
+        check skipToClose(tok, "]");
+        t = tok.current();
     }
-    check tok.advance();
-    t = tok.current();
-    match t {
-        "." => {
-            // semantically invalid
-            return PREDICT_EXPR;
+    return t != () && t != ".";
+}
+
+function skipToClose(Tokenizer tok, string? close) returns err:Syntax? {
+    while true {
+        Token? t = tok.current();
+        if t == () || t == ";" && close != "|}" {
```

Use parentheses to make precedence explicit here.

In the case of `;`, I think you should return an error. This function is declared as returning an error, but doesn't do so.